File: Python/105.Construct Binary Tree from Preorder and Inorder Traversal.py
#根据二叉树的前序遍历和中序遍历复原二叉树
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        if preorder==[]:
            return None
        return self.buildSubTree(preorder,inorder,0,len(preorder)-1,0,len(inorder)-1)

    def buildSubTree(self,preorder,inorder,l,r,ll,rr):
        if l>r or ll>rr:
            return None
        node = TreeNode(preorder[l])
        i = ll
        while i<=rr:
            if inorder[i]==preorder[l]:
                break
            i += 1
        length1 = i-ll-1    ##表示分为左子树的长度
        length2 = rr-i-1    ##表示分为右子树的长度
        node.left = self.buildSubTree(preorder,inorder,l+1,i-ll+l,ll,i-1)
        node.right = self.buildSubTree(preorder,inorder,r-rr+i+1,r,i+1,rr)
        return node

class Ssolution(object):
    def inorderTraversal(self, root):
        self.result = []
        self.inorder(root)
        return self.result

    def inorder(self,p):
        if p!=None:
            self.inorder(p.left) if p.left!=None else None
            self.result.append(p.val)
            self.inorder(p.right) if p.right!=None else None

ss = Ssolution()
s = Solution()
root = s.buildTree([1,2,4,6,3,7,5],[4,2,6,1,7,3,5])
print(ss.inorderTraversal(root))

File: C++/Array/105.Construct Binary Tree from Preorder and Inorder Traversal.cpp
/*
给出二叉树的前序和中序遍历，求出二叉树。
前序：根节点，左子树，右子树
中序：左子树，根节点，右子树
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<int> mypreorder;
	vector<int> myinorder;
public:
	TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
		if (preorder.size() == 0)	return NULL;
		mypreorder = preorder, myinorder = inorder;
		return buildSubTree(0, preorder.size() - 1, 0, inorder.size() - 1);
	}
	TreeNode* buildSubTree(int preleft, int preright,int inleft, int inright )
	{
		if (preleft>preright)	return NULL;
		TreeNode *root = new TreeNode(mypreorder[preleft]);
		for (int i = inleft; i <= inright; ++i)
		{
			if (myinorder[i] == root->val)
			{
				root->left = buildSubTree(preleft + 1, preleft+i-inleft, inleft, i - 1);
				root->right = buildSubTree(preright-(inright-i-1), preright, i + 1, inright);
				return root;
			}
		}
		return NULL;
	}
};

void postorder(TreeNode *root)
{
	if (root)
	{
		postorder(root->left);
		postorder(root->right);
		cout << root->val << endl;
	}
}

int main()
{
	Solution s;
	vector<int> preorder{ 1,2,3 };
	vector<int> inorder{ 2,3,1 };
	TreeNode *tree = s.buildTree(preorder, inorder);
	postorder(tree);
}

